python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,491 @@
+
+        return next_group
+
+    # put a job in the runner queue and assign a runner thread to do some work
+    def assignRunner(self, job):
+        self.runner_queue.put(job)
+        return self.runner_pool.apply_async(self.jobRunner, (self.runner_queue,))
+
+    # put a job in the status queue and assign a status thread to do some work
+    def assignStatus(self, job):
+        self.status_queue.put(job)
+        return self.status_pool.apply_async(self.statusRunner, (self.status_queue,))
+
+    # Status processing (hand a tester's status back to the TestHarness to have some status printed to the screen)
+    def statusRunner(self, queue):
+        job = queue.get()
+        self.harness.handleTestStatus(self.scheduled_tester_data[job])
+
+    # Runner processing (call derived run method using a thread)
+    def jobRunner(self, queue):
+        job = queue.get()
+        tester_dag = self.scheduled_tester_data[job].getDAG()
+        tester = self.scheduled_tester_data[job].getTester()
+
+        # Check if we have available slots
+        if self.checkAvailableSlots(tester):
+            can_run = True
+            with self.thread_lock:
+                # Add this job to our active list
+                self.active_jobs.add(job)
+
+                # Record the amount of slots we're about to consume
+                self.workers_in_use += tester.getProcs(self.options)
+
+        # We can not run just yet (slots full)
+        else:
+            sleep(0.03)
+            self.assignRunner(job)
+            return
+
+        # Get a timer tuple of started thread timers
+        my_timers = self.handleTimers(start_timers=self.scheduled_tester_data[job])
+
+        # Call derived run method outside of the thread_lock (run is blocking)
+        self.run(self.scheduled_tester_data[job])
+
+        # The test finished, so stop any timers we received when starting them
+        self.handleTimers(stop_timers=my_timers)
+
+        # run should never set a pending status. If it did, something is wrong with the derived scheduler
+        # TODO: create a unit test for this
+        if tester.isPending():
+            raise SchedulerError('Derived Scheduler can not return a pending status!')
+
+        # Add this job to the status queue for finished work to be printed to the screen
+        self.assignStatus(job)
+
+        # This job is finished, prepare to launch other avialable jobs out of the DAG
+        with self.thread_lock:
+            # Recover our available slots
+            self.workers_in_use = max(0, self.workers_in_use - tester.getProcs(self.options))
+
+            # Remove this job from our active set
+            self.active_jobs.remove(job)
+
+        # Get the next group of jobs we can run
+        next_job_group = self.getNextJobGroup(self.scheduled_tester_data[job])
+
+        # Launch these next set of jobs
+        for next_job in next_job_group:
+            self.assignRunner(next_job)
```

Comment:
Yeah - this is all closer then sending in the DAG. The items in the queue should be Testers (or more accurately TesterDatas) of course in Python under the hood that means references to the actual items which sit in the DAG.